### ISSUE TITLE
Cannot use Microsoft.Identity.Client 1.0.304142221-alpha for ASP.NET Core 1.1

### ISSUE THREAD

**Issue by frankchen76** [AzureAD/microsoft-authentication-library-for-dotnet#261]
trying to use MSAL 1.0.304142221-alpha in my ASP.NET Core 1.1 to connect to the latest AAD V2 endpoint. I wasn't able to OpenID Connect to work. 
The following is my code: 
Startup.cs
`    public class Startup
    {
        public Startup(IHostingEnvironment env)
        {
            var builder = new ConfigurationBuilder()
                .SetBasePath(env.ContentRootPath)
                .AddJsonFile("appsettings.json", optional: true, reloadOnChange: true)
                .AddJsonFile($"appsettings.{env.EnvironmentName}.json", optional: true)
                .AddEnvironmentVariables();
            Configuration = builder.Build();
        }

        public IConfigurationRoot Configuration { get; }

        // This method gets called by the runtime. Use this method to add services to the container.
        public void ConfigureServices(IServiceCollection services)
        {
            AadConfig aadConfig = new AadConfig();
            Configuration.GetSection(AadConfig.SECTION_NAME).Bind(aadConfig);

            AadConfig.InitConfig(aadConfig);
            services.Configure<AadConfig>(Configuration.GetSection(AadConfig.SECTION_NAME));
            // Add framework services.
            services.AddMvc();

            //add Session
            services.AddSession();

            services.TryAddSingleton<IHttpContextAccessor, HttpContextAccessor>();

        }

        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
        public void Configure(IApplicationBuilder app, IHostingEnvironment env, ILoggerFactory loggerFactory)
        {
            loggerFactory.AddConsole(Configuration.GetSection("Logging"));
            loggerFactory.AddDebug();

            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
                app.UseBrowserLink();
            }
            else
            {
                app.UseExceptionHandler("/Home/Error");
            }

            app.UseSession();
            IHttpContextAccessor iHttpContext = app.ApplicationServices.GetService<IHttpContextAccessor>();

            IOptions<AadConfig> aadConfig = app.ApplicationServices.GetService<IOptions<AadConfig>>();

            app.UseStaticFiles();

            // Configure the OWIN pipeline to use cookie auth.
            app.UseCookieAuthentication(new CookieAuthenticationOptions());

            // Configure the OWIN pipeline to use OpenID Connect auth.
            
            app.UseOpenIdConnectAuthentication(new OpenIdConnectOptions
            {
                ClientId = Configuration["AzureAD:AppId"],
                Authority = Configuration["AzureAd:AadInstance"], //string.Format(CultureInfo.InvariantCulture, Configuration["AzureAd:AadInstance"], "common", "/v2.0"),
                ResponseType = OpenIdConnectResponseType.CodeIdToken,
                PostLogoutRedirectUri = Configuration["AzureAd:PostLogoutRedirectUri"],
                CallbackPath= Configuration["AzureAd:CallbackUrl"],
                //Resource= Configuration["AzureAd:GraphScopes"],
                //Scope = "openid email profile offline_access " + Configuration["AzureAD:GraphScopes"],
                Events = new OpenIdConnectEvents
                {
                    OnRemoteFailure = RemoteFailure,
                    OnTokenValidated = TokenValidated,
                    OnAuthorizationCodeReceived = async (context) =>
                     {
                         string redirectUri = aadConfig.Value.RedirectUri;
                         string appId = aadConfig.Value.AppId;
                         string auth = String.Format("https://login.microsoftonline.com/{0}/v2.0", appId);
                         string appSecret = aadConfig.Value.AppSecret;
                         string[] scopes = aadConfig.Value.GraphScopes.Split(new char[] { ' ' });

                         var code = context.ProtocolMessage.Code;
                        //string signedInUserID= iHttpContext.HttpContext.User.FindFirst(ClaimTypes.NameIdentifier).Value;
                        string signedInUserID = context.JwtSecurityToken.Subject;
                        //string signedInUserID = context.AuthenticationTicket.Identity.FindFirst(ClaimTypes.NameIdentifier).Value;
                        ConfidentialClientApplication cca = new ConfidentialClientApplication(
                             auth,
                             appId,
                             redirectUri,
                             new ClientCredential(appSecret),
                             new SessionTokenCache(signedInUserID, iHttpContext.HttpContext));

                         AuthenticationResult result = await cca.AcquireTokenByAuthorizationCodeAsync(scopes, code);
                         string token = result.Token;
                         context.HandleCodeRedemption(result.Token, result.IdToken);
                     },
                },
                SignInScheme= CookieAuthenticationDefaults.AuthenticationScheme,
                TokenValidationParameters = new TokenValidationParameters
                {
                    // Instead of using the default validation (validating against
                    // a single issuer value, as we do in line of business apps), 
                    // we inject our own multitenant validation logic
                    ValidateIssuer = false,

                    NameClaimType = "name"
                }
                
            });
            app.UseMvc(routes =>
            {
                routes.MapRoute(
                    name: "default",
                    template: "{controller=Home}/{action=Index}/{id?}");
            });
        }

        private Task TokenValidated(TokenValidatedContext context)
        {
            /* ---------------------
            // Replace this with your logic to validate the issuer/tenant
               ---------------------       
            // Retriever caller data from the incoming principal
            string issuer = context.SecurityToken.Issuer;
            string subject = context.SecurityToken.Subject;
            string tenantID = context.Ticket.Principal.FindFirst("http://schemas.microsoft.com/identity/claims/tenantid").Value;

            // Build a dictionary of approved tenants
            IEnumerable<string> approvedTenantIds = new List<string>
            {
                "<Your tenantID>",
                "9188040d-6c67-4c5b-b112-36a304b66dad" // MSA Tenant
            };

            if (!approvedTenantIds.Contains(tenantID))
                throw new SecurityTokenValidationException();
              --------------------- */

            return Task.FromResult(0);
        }

        // Handle sign-in errors differently than generic errors.
        private Task RemoteFailure(FailureContext context)
        {
            context.HandleResponse();
            context.Response.Redirect("/Home/Error?message=" + context.Failure.Message);
            return Task.FromResult(0);
        }
    }`
the problem is I don't know whether I should listen OnAuthorizationCodeReceived event. Based on [ASP.NET 4.6 sample](https://github.com/microsoftgraph/aspnet-connect-sample), we need to listen this event and store the access code to the session by calling AcquiredTokenByAuthorizationCode(). However, I got error message to say "AADSTS70000: The access grant was obtained for a different tenant" if I used the above code. 
I also tried not listen this event and set ResponseType = OpenIdConnectResponseType.IdToken, my ASP.NET Core MVC 1.1 application can be authenticated. however, I cannot call ConfidentialClientApplication.AcquireTokenSilentAsync() get access token to access graph api.

**Comment by kpanwar** [comment]
This is an error coming from the server. MSAL simply returns what it got back from server.

**Comment by frankchen76** [comment]
I finally found out the reason, I use the different authority url in my code which caused the error. after I changed Authority URL to "https://login.microsoftonline.com/common/v2.0" for OpenID Connect and MSAL, it works fine. 
a few catch: 
- I have to set `ResponseType = OpenIdConnectResponseType.CodeIdToken `in OpenID Connect so that OnAuthorizationCodeReceived will be fired. 
- To get auth code, I have to use `var code = context.ProtocolMessage.Code;` instead of `var code = context.Code;` for MSAL
- For Signinuserid, I can get from JwtSecurityToken.Subject like `string signedInUserID = context.JwtSecurityToken.Subject`
- I need to call `context.HandleCodeRedemption(result.Token, result.IdToken);` to use access token.
